### PR TITLE
Show yanked gem in search filters

### PR DESCRIPTION
**Pull request by vachhanihpavan**
Changes Done:

1. Show filter `YANKED (1)` if a gem with exact matching name exists.
2. Show yanked badge

:computer: Screenshots
search

All versions of gem `a` is yanked. So, it won't appear in search results. 
A filter will appear if a gem with exact name matches.
Refs:
[1](https://github.com/rubygems/rubygems.org/pull/2262#issuecomment-599189758)
[2](https://github.com/rubygems/rubygems.org/pull/2262#issuecomment-620031464)

yanked-filter
On clicking the YANKED filter.

Rework on #2110

### CODE REVIEW

File: app/helpers/searches_helper.rb [hunk 1]
```diff
@@ -26,4 +26,11 @@ def aggregation_count(aggregration, duration, buckets_pos)
     update_info = (duration == 30.days ? t("searches.show.month_update", count: count) : t("searches.show.week_update", count: count))
     link_to update_info, path, class: "t-link--black"
   end
+
+  def aggregation_yanked(yanked_gem)
+    return unless yanked_gem
+
+    path = search_path(params: { query: params[:query], yanked: true })
+    link_to t("versions.version.yanked") + " (1)", path, class: "t-link--black"
```
Comment (vachhanihpavan) [inline app/helpers/searches_helper.rb:34]: Is this fine?
`t("versions.version.yanked") + " (1)"`
Or better to make a new key? Though count will always be 1

Comment (vachhanihpavan) [reply app/helpers/searches_helper.rb]: Added a new key with i18n value

File: app/views/searches/show.html.erb [hunk 2]
```diff
@@ -6,28 +6,44 @@
   </div>
 <% end %>
 <%= link_to t("advanced_search"), advanced_search_path, class: "t-link--gray t-link--has-arrow" %>
-<% if @gems %>
+<% if params[:yanked] %>
```
Comment (sonalkr132) [inline app/views/searches/show.html.erb:9]: this condition should exist in controller
https://stackoverflow.com/questions/15823695/best-practice-passing-instance-variables-or-using-params-in-ruby-on-rails-view

File: app/views/searches/show.html.erb [hunk 3]
```diff
@@ -6,28 +6,44 @@
...
-        end.to_sentence(last_word_connector: ' or ').html_safe %>?
-      </p>
-    </div>
-  <% end %>
+    <%= render partial: 'aggregations', locals: { gems: @gems } %>
 
-  <% if @gems.present? %>
-    <%= render partial: 'rubygems/rubygem', collection: @gems %>
-    <%= paginate @gems %>
+    <% if suggestions = es_suggestions(@gems)  %>
```
Comment (vachhanihpavan) [inline app/views/searches/show.html.erb:33]: @sonalkr132 I think this also isn't a good way. 
`suggestions = es_suggestions(@gems)` returns nil if `es_suggestions(@gems)` is nil.
But assignment in if condition is not cool.

We can change it to 
```
<% suggestions = es_suggestions(@gems) %>
<% if suggestions %>
```

Comment (sonalkr132) [reply app/views/searches/show.html.erb:33]: yes, although rubocop doesn't support erb.. we try to follow all of [style guides](https://github.com/rubocop-hq/ruby-style-guide#safe-assignment-in-condition) they list.

File: app/views/searches/show.html.erb [hunk 4]
```diff
@@ -6,28 +6,44 @@
...
+      </div>
+    <% end %>
+
+    <% if @gems.present? %>
+      <%= render partial: 'rubygems/rubygem', collection: @gems %>
+      <%= paginate @gems %>
+    <% end %>
   <% end %>
-<% end %>
+<% end %>
```
Comment (sonalkr132) [inline app/views/searches/show.html.erb:49]: new line, please.

File: app/controllers/searches_controller.rb [hunk 5]
```diff
@@ -5,6 +5,10 @@ def show
     return unless params[:query]&.is_a?(String)
     @error_msg, @gems = ElasticSearcher.new(params[:query], page: @page).search
     limit_total_count if @gems.total_count > Gemcutter::SEARCH_MAX_PAGES * Rubygem.default_per_page
+
+    exact_match = Rubygem.name_is(params[:query]).first
+    @yanked_gem = exact_match unless exact_match&.indexed_versions?
+    @yanked_filter = params[:yanked]
```
Comment (sonalkr132) [inline app/controllers/searches_controller.rb:11]: maybe check if params is  `true` and set var to boolean?

Comment (sonalkr132) [reply app/controllers/searches_controller.rb:11]: we want to make sure we aren't using any user input data directly in view. they can send all kinds of garbage.

Comment (vachhanihpavan) [reply app/controllers/searches_controller.rb]: Done! :+1:

File: test/functional/searches_controller_test.rb [hunk 6]
```diff
@@ -123,6 +123,27 @@ class SearchesControllerTest < ActionController::TestCase
...
 
+  context "on GET to show with search parameters with yanked gems" do
+    setup do
+      @sinatra = create(:rubygem, name: "sinatra")
+      @sinatra_redux = create(:rubygem, name: "sinatra-redux")
+      create(:version, rubygem: @sinatra)
+      create(:version, :yanked, rubygem: @sinatra_redux)
+      import_and_refresh
+      @request.cookies["new_search"] = "true"
+      get :show, params: { query: @sinatra_redux.name.to_s }
```
Comment (sonalkr132) [inline test/functional/searches_controller_test.rb:134]: please updated this to use `yanked` param.
`@request.cookies["new_search"] = "true"` is not needed. remove it from your setup and maybe send a new PR updating old these where this is still being used.

**Comment by sonalkr132** [comment]
Thank you for working on this. :green_heart:

**Comment by ioquatix** [comment]
Thanks for implementing this, it will be really helpful.

**Pull request merged**